ateam_ai/src/trajectory_generation/ilqr_problem.hpp

```diff
@@ -0,0 +1,290 @@
+// https://studywolf.wordpress.com/2016/02/03/the-iterative-linear-quadratic-regulator-method/
+// X = num states, U = num inputs, T = num timesteps
+template<int X, int U, int T>
+class iLQRProblem
+{
+public:
+  using State = Eigen::Matrix<double, X, 1>;
+  using Input = Eigen::Matrix<double, U, 1>;
+  using Jacobian = Eigen::Matrix<double, X, X + U>;
+  using Cost = double;
+  using Gradiant = Eigen::Matrix<double, X + U, 1>;
+  using Hessian = Eigen::Matrix<double, X + U, X + U>;
+  using Time = int;
+
+  using Trajectory = std::array<State, T>;
+  using Actions = std::array<Input, T>;
+
+  using Feedforwards = std::array<Input, T>;
+  using Feedbacks = std::array<Eigen::Matrix<double, U, X>, T>;
+
+  // Assume second derivative of dynamics is 0 so we don't have do tensor math
+  virtual State dynamics(const State & x_t1, const Input & u_t1, Time t) = 0;
+  virtual Jacobian jacobian(const State & x, const Input & u, Time t)
+  {
+    Jacobian out;
+
+    State d = dynamics(x, u, t);
+    for (std::size_t i = 0; i < X; i++) {
+      State x_eps = x;
+      x_eps(i) += eps;
+      out.col(i) = (dynamics(x_eps, u, t) - d) / eps;
+    }
+
+    for (std::size_t i = 0; i < U; i++) {
+      Input u_eps = u;
+      u_eps(i) += eps;
+      out.col(i + X) = (dynamics(x, u_eps, t) - d) / eps;
+    }
+
+    return out;
+  }
+
+  virtual Cost cost(const State & x, const Input & u, Time t) = 0;
+  virtual Gradiant gradiant(const State & x, const Input & u, Time t)
+  {
+    Gradiant out;
+
+    for (std::size_t i = 0; i < X; i++) {
+      State x_eps_p = x;
+      State x_eps_n = x;
+      x_eps_p(i) += eps;
+      x_eps_n(i) -= eps;
+      out(i) = (cost(x_eps_p, u, t) - cost(x_eps_n, u, t)) / (2 * eps);
+    }
+
+    for (std::size_t i = 0; i < U; i++) {
+      Input u_eps_p = u;
+      Input u_eps_n = u;
+      u_eps_p(i) += eps;
+      u_eps_n(i) -= eps;
+      out(i + X) = (cost(x, u_eps_p, t) - cost(x, u_eps_n, t)) / (2 * eps);
+    }
+
+    return out;
+  }
+
+  virtual Hessian hessian(const State & x, const Input & u, Time t, const Gradiant & g)
+  {
+    Hessian out;
+
+    for (std::size_t i = 0; i < X; i++) {
+      State x_eps = x;
+      x_eps(i) += eps;
+      out.col(i) = (gradiant(x_eps, u, t) - g) / eps;
+    }
+
+    for (std::size_t i = 0; i < U; i++) {
+      Input u_eps = u;
+      u_eps(i) += eps;
+      out.col(i + X) = (gradiant(x, u_eps, t) - g) / eps;
+    }
+
+    return out;
+  }
+
+  Trajectory calculate(const State & initial_state)
+  {
+    forward_rollout(initial_state);
+    backward_pass();
+
+    return trajectory;
+  }
+
+private:
+  using Jacobians = std::array<Jacobian, T>;
+  using Hessians = std::array<Hessian, T>;
+  using Gradiants = std::array<Gradiant, T>;
+
+
+  void forward_rollout(const State & initial_state)
+  {
+    // Step 1: Forward Rollout
+    trajectory.front() = initial_state;
+    actions.front().setZero();
+    overall_cost = cost(trajectory.at(0), actions.at(0), 0);
```

Comment:
nit: consider `std::array<>::front()`
```suggestion
    overall_cost = cost(trajectory.front(), actions.front(), 0);
```